src/drain.rs

```diff
@@ -0,0 +1,80 @@
+use std::future::Future;
+use std::time::Duration;
+use tokio::sync::watch;
+use tracing::{debug, info, warn};
+
+#[derive(Debug)]
+pub struct DrainTrigger(drain_internal::Signal);
+
+impl DrainTrigger {
+    /// start_drain_and_wait initiates a draining sequence. The future will not complete until the drain
+    /// is complete (all outstanding DrainWatchers are dropped).
+    pub async fn start_drain_and_wait(self) {
+        self.0.drain().await
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct DrainWatcher(drain_internal::Watch);
+
+impl DrainWatcher {
+    /// wait_for_drain will return once a drain has been initiated.
+    /// The drain will not complete until the returned DrainBlocker is dropped
+    pub async fn wait_for_drain(self) -> DrainBlocker {
+        DrainBlocker(self.0.signaled().await)
+    }
+}
+
+#[allow(dead_code)]
+/// DrainBlocker provides a token that must be dropped to unblock the drain.
+pub struct DrainBlocker(drain_internal::ReleaseShutdown);
+
+/// New constructs a new pair for draining
+/// * DrainTrigger can be used to start a draining sequence and wait for it to complete.
+/// * DrainWatcher should be held by anything that wants to participate in the draining. This can be cloned,
+///   and a drain will not complete until all outstanding DrainWatchers are dropped.
+pub fn new() -> (DrainTrigger, DrainWatcher) {
+    let (tx, rx) = drain_internal::channel();
+    (DrainTrigger(tx), DrainWatcher(rx))
+}
+
```

Comment:
Why do we need `DrainBlocker` and friends? `tokio::sync::watch` supports all the same semantics and has `closed().await` which already blocks until all listeners are dropped? We can just use that, or am I missing something?

Reply:
Here is my thinking:

* Having nicely named things, with clear semantics (i.e. the ONLY thing you can do with a DrainTrigger is `start_drain_and_wait`) is a good idea, regardless of the underlying implementation. If I just have a `Watch`, its not clear what its for, and how to use it.
* If we abstract it ,we can change the underlying implementation to directly use `tokio::*` instead of `drain::*`, if there is benefit to do so

Reply:
> If we abstract it ,we can change the underlying implementation to directly use tokio::* instead of drain::*, if there is benefit to do so

Yeah my point is I think the abstraction doesn't actually provide any benefit (other than the type name alias) if we just use `tokio::sync::watch`. Additionally, doing that lets us drop a Cargo dep we really don't need, that (IIRC) we are now only using here - and dropping deps is always good, especially in `ztunnel`.

I am fine wrapping `tokio::sync::watch` with this wrapper if we want to keep the type alias, but I don't want to keep `drain` AND the type alias if regular old `tokio::sync::watch` (which is already one of our deps) will do the job.

Reply:
Here is what I want:
* A wrapper with reasonable names and exactly one method. Watch has bad names (since they are general) and too many methods (since its general) making it hard to use
* To not spend time _right now_ migrating off `drain` since it works fine.

If someone wants to swap out the underlying implement of `drain`, now its trivial to do so -- edit the tiny new package. I am not actually convinced its better (maybe it is, just haven't looked into it), though. I do not think we can SIMPLY stick a watch in there, probably we will end up with 2, and then we have just re-inventing `drain`.Which is fine if we have reasons to, but its super low priority, and we need this PR anyways.